glean-core/csharp/Glean/Metrics/PingType.cs

```diff
@@ -0,0 +1,78 @@
+﻿// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+using Mozilla.Glean.FFI;
+using System;
+using static Mozilla.Glean.Glean;
+
+namespace Mozilla.Glean.Private
+{
+    /// <summary>
+    ///  An enum with no values for convenient use as the default set of reason codes.
+    /// </summary>
+    public enum NoReasonCodes : int
+    {
+        value
+    }
+
+    /// <summary>
+    ///  The base class of all PingTypes with just enough to track their registration, so
+    ///  we can create a heterogeneous collection of ping types.
+    /// </summary>
+    public class PingTypeBase
+    {
+        internal string name;
+        internal LibGleanFFI.PingTypeHandle handle;
+
+        internal PingTypeBase() { }
+    }
+
+    /// <summary>
+    ///  This implements the developer facing API for custom pings.
+    ///  Instances of this class type are automatically generated by the parsers at build time.
+    ///  The Ping API only exposes the [send] method, which schedules a ping for sending.
+    /// </summary>
+    public class PingType<NoReasonCodes> : PingTypeBase
```

Comment:
```suggestion
    public class PingType<ReasonCodesEnum> : PingTypeBase where ReasonCodesEnum: System.Enum, struct
```

The intent here is for `PingType` to declare the type for the Generic class and also add a constraint to it: we only want enums to be valid generics arguments.

**Note**: `struct` is needed in order to allow for passing a nullable enum to `Submit`